File: m3u8gen/src/m3u8.java
public class m3u8 {
  // Live Playlist
  public static String GetLivePlaylist(Integer target, Integer count, Integer[] seqs, Integer[] infs, String[] urls) {
    // check EXT-X-TARGETDURATION
    if (target < 1) {
      System.out.println(String.format("invalid TARGETDURATION: %d", target));
      return "";
    }

    // check array length
    if (count < 1) {
      System.out.println(String.format("invalid count: %d", count));
      return "";
    }
    if (count != seqs.length || count != infs.length || count != urls.length) {
      System.out.println(String.format("invalid length: count=%d, seqs=%d, infs=%d, urls=%d", count, seqs.length, infs.length, urls.length));
      return "";
    }

    // check EXT-X-MEDIA-SEQUENCE
    if (seqs[0] < 0) {
      System.out.println(String.format("invalid SEQUENCE: %d", seqs[0]));
      return "";
    }

    String playlist = "#EXTM3U\r\n";
    playlist += "#EXT-X-VERSION:1\r\n";
    playlist += "#EXT-X-TARGETDURATION:" + String.valueOf(target) + "\r\n";
    playlist += "#EXT-X-MEDIA-SEQUENCE:" + String.valueOf(seqs[0]) + "\r\n";

    for (int i = 0; i < count; i++) {
      // check array element
      if (seqs[i] < 0 || infs[i] < 1 || urls[i].isEmpty()) {
        System.out.println(String.format("invalid element: seq=%d, inf=%d, url=%s", seqs[i], infs[i], urls[i]));
        return "";
      }

      // check EXT-X-DISCONTINUITY
      if (i > 0) {
        if (seqs[i] <= seqs[i-1]) {
          System.out.println(String.format("invalid DISCONTINUITY between [%d, %d]", seqs[i-1], seqs[i]));
          return "";
        }

        for (int d = seqs[i-1]+1; d < seqs[i]; d++) {
          System.out.println(String.format("DISCONTINUITY [%d] between [%d, %d]", d, seqs[i-1], seqs[i]));
          playlist += "#EXT-X-DISCONTINUITY\r\n";
        }
      }

      playlist += "#EXTINF:" + String.valueOf(infs[i]) + ",\r\n";
      playlist += urls[i] + "\r\n";
    }

    return playlist;
  }

  // VOD Playlist
  public static String GetVoDPlaylist(Integer target, Integer seq, Integer count, Integer[] infs, String[] urls) {
    // check EXT-X-TARGETDURATION
    if (target < 1) {
      System.out.println(String.format("invalid TARGETDURATION: %d", target));
      return "";
    }

    // check EXT-X-MEDIA-SEQUENCE
    if (seq < 0) {
      System.out.println(String.format("invalid SEQUENCE: %d", seq));
      return "";
    }

    // check array length
    if (count < 1) {
      System.out.println(String.format("invalid count: %d", count));
      return "";
    }
    if (urls.length != count || infs.length != count) {
      System.out.println(String.format("invalid length: count=%d, infs=%d, urls=%d", count, infs.length, urls.length));
      return "";
    }

    String playlist = "#EXTM3U\r\n";
    playlist += "#EXT-X-VERSION:1\r\n";
    playlist += "#EXT-X-PLAYLIST-TYPE:VOD\r\n";
    playlist += "#EXT-X-TARGETDURATION:" + String.valueOf(target) + "\r\n";
    playlist += "#EXT-X-MEDIA-SEQUENCE:" + String.valueOf(seq) + "\r\n";

    for (int i = 0; i < count; i++) {
      // check array element
      if (infs[i] < 1 || urls[i].isEmpty()) {
        System.out.println(String.format("invalid element: inf=%d, url=%s", infs[i], urls[i]));
        return "";
      }

      playlist += "#EXTINF:" + String.valueOf(infs[i]) + ",\r\n";
      playlist += urls[i] + "\r\n";
    }

    playlist += "#EXT-X-ENDLIST\r\n";
    return playlist;
  }

  // EVENT Playlist
  public static String GetEventPlaylist(Integer target, Integer seq, Integer count, Integer[] infs, String[] urls, Boolean head, Boolean end) {
    // check EXT-X-TARGETDURATION
    if (target < 1) {
      System.out.println(String.format("invalid TARGETDURATION: %d", target));
      return "";
    }

    // check EXT-X-MEDIA-SEQUENCE
    if (seq < 0) {
      System.out.println(String.format("invalid SEQUENCE: %d", seq));
      return "";
    }

    // check array length
    if (count < 1) {
      System.out.println(String.format("invalid count: %d", count));
      return "";
    }
    if (urls.length != count || infs.length != count) {
      System.out.println(String.format("invalid length: count=%d, infs=%d, urls=%d", count, infs.length, urls.length));
      return "";
    }

    String playlist = "";
    if (head) {
      playlist += "#EXTM3U\r\n";
      playlist += "#EXT-X-VERSION:1\r\n";
      playlist += "#EXT-X-PLAYLIST-TYPE:EVENT\r\n";
      playlist += "#EXT-X-TARGETDURATION:" + String.valueOf(target) + "\r\n";
      playlist += "#EXT-X-MEDIA-SEQUENCE:" + String.valueOf(seq) + "\r\n";
    }

    for (int i = 0; i < count; i++) {
      // check array element
      if (infs[i] < 1 || urls[i].isEmpty()) {
        System.out.println(String.format("invalid element: inf=%d, url=%s", infs[i], urls[i]));
        return "";
      }

      playlist += "#EXTINF:" + String.valueOf(infs[i]) + ",\r\n";
      playlist += urls[i] + "\r\n";
    }

    if (end) {
      playlist += "#EXT-X-ENDLIST\r\n";
    }
    return playlist;
  }
}

File: m3u8.js
const { TS_CLIP_DURATION, getSeq, getUrl, getKeysLatest, getKeysBetween } = require('./s3_client');

async function getLivePlaylist(cameraId, nowTime, lastSeq) {
  let keys = await getKeysLatest(cameraId, nowTime);
  if (keys.length < 1) {
    console.log(`invalid count: ${keys.length}`);
    return {m3u8: null, seq: -1};
  }

  const first_seq = await getSeq(keys[0]);
  if (first_seq < 0) {
    console.log(`invalid SEQUENCE: ${first_seq}`);
    return {m3u8: null, seq: -1};
  }

  let playlist = '#EXTM3U\r\n';
  playlist += "#EXT-X-VERSION:1\r\n";
  playlist += '#EXT-X-TARGETDURATION:' + TS_CLIP_DURATION + '\r\n';
  playlist += '#EXT-X-MEDIA-SEQUENCE:' + first_seq + '\r\n';

  let last_seq = -1;
  let last_key;
  for (let key of keys) {
    const seq = await getSeq(key);
    if (seq < 0) {
      console.log(`invalid element: seq=${seq}, key=${key}`);
      return {m3u8: null, seq: -1};
    }

    if (last_seq >= 0) {
      if (seq <= last_seq) {
        console.log(`invalid DISCONTINUITY between [${last_seq}, ${seq}]`);
        return {m3u8: null, seq: -1};
      }
      for (let d = last_seq + 1; d < seq; d++) {
        console.log(`DISCONTINUITY [${d}] between [${last_seq}, ${seq}]`);
        playlist += "#EXT-X-DISCONTINUITY\r\n";
      }
    }

    last_seq = seq;
    last_key = key;

    const url = await getUrl(key);
    playlist += '#EXTINF:' + TS_CLIP_DURATION + ',\r\n';
    playlist += url + '\r\n';
  }

  return {m3u8: playlist, seq: first_seq};
}

async function getVoDPlaylist(cameraId, startTime, endTime) {
  let keys = await getKeysBetween(cameraId, startTime, endTime);
  if (keys.length < 1) {
    console.log(`invalid count: ${keys.length}`);
    return {m3u8: null, seq: -1};
  }

  const first_seq = await getSeq(keys[0]);
  if (first_seq < 0) {
    console.log(`invalid SEQUENCE: ${first_seq}`);
    return {m3u8: null, seq: -1};
  }

  let playlist = '#EXTM3U\r\n';
  playlist += "#EXT-X-VERSION:1\r\n";
  playlist += "#EXT-X-PLAYLIST-TYPE:VOD\r\n";
  playlist += '#EXT-X-TARGETDURATION:' + TS_CLIP_DURATION + '\r\n';
  playlist += '#EXT-X-MEDIA-SEQUENCE:' + first_seq + '\r\n';

  for (let key of keys) {
    const url = await getUrl(key);
    playlist += '#EXTINF:' + TS_CLIP_DURATION + ',\r\n';
    playlist += url + '\r\n';
  }

  playlist += '#EXT-X-ENDLIST\r\n';
  return {m3u8: playlist, seq: first_seq};
}

async function getEventPlaylist(cameraId, startTime, endTime, lastM3U8, lastSeq, appendNum) {
  let keys = await getKeysBetween(cameraId, startTime, endTime);
  if (keys.length < 1) {
    console.log(`invalid count: ${keys.length}`);
    return {m3u8: null, seq: -1};
  }

  const first_seq = await getSeq(keys[0]);
  if (first_seq < 0) {
    console.log(`invalid SEQUENCE: ${first_seq}`);
    return {m3u8: null, seq: -1};
  }

  let playlist;
  if (lastM3U8) {
    playlist = lastM3U8;
  } else {
    playlist = '#EXTM3U\r\n';
    playlist += "#EXT-X-VERSION:1\r\n";
    playlist += "#EXT-X-PLAYLIST-TYPE:EVENT\r\n";
    playlist += '#EXT-X-TARGETDURATION:' + TS_CLIP_DURATION + '\r\n';
    playlist += '#EXT-X-MEDIA-SEQUENCE:' + first_seq + '\r\n';
  }

  var last_seq = -1;
  var total = 0;
  var count = 0;
  for (let key of keys) {
    last_seq = await getSeq(key);
    if (last_seq < 0) {
      console.log(`invalid SEQUENCE: ${last_seq}`);
      return {m3u8: null, seq: -1};
    }

    total++;
    if (last_seq <= lastSeq) {
      continue;
    }

    const url = await getUrl(key);
    playlist += '#EXTINF:' + TS_CLIP_DURATION + ',\r\n';
    playlist += url + '\r\n';

    count++;
    if (count >= appendNum) {
      break;
    }
  }

  if (last_seq < lastSeq) {
    console.log(`incomplete playlist: ${last_seq} < ${lastSeq}`);
    return {m3u8: null, seq: -1};
  }

  if (total == keys.length && last_seq != lastSeq) {
    playlist += '#EXT-X-ENDLIST\r\n';
  }
  return {m3u8: playlist, seq: last_seq};
}

module.exports = {
  getLivePlaylist,
  getVoDPlaylist,
  getEventPlaylist
}
